File: internal/server/protocol.go
package server

// WebSocket wire shapes (SPEC §8). web/src/api/protocol.ts mirrors these.

import (
	"encoding/json"

	"github.com/jeffbstewart/unconf/internal/domain"
	"github.com/jeffbstewart/unconf/internal/store"
)

// clientFrame is any client → server frame.
type clientFrame struct {
	Type    string          `json:"type"` // "cmd" | "ping"
	CmdID   string          `json:"cmdId"`
	Cmd     string          `json:"cmd"`
	Payload json.RawMessage `json:"payload"`
}

type userJSON struct {
	ID   string      `json:"id"`
	Name string      `json:"name"`
	Role domain.Role `json:"role"`
}

func toUserJSON(u store.User) userJSON { return userJSON{u.ID, u.Name, u.Role} }

type noteJSON struct {
	ID        string        `json:"id"`
	Title     string        `json:"title"`
	BodyMD    string        `json:"bodyMd"`
	AuthorID  string        `json:"authorId"`
	X         float64       `json:"x"`
	Y         float64       `json:"y"`
	Color     string        `json:"color"`
	RegionID  *string       `json:"regionId"`
	VoteTotal int           `json:"voteTotal"`
	Voters    []string      `json:"voters"` // user ids; votes are public (SPEC §11)
	Voted     bool          `json:"voted"`  // this user voted for it
	Starred   bool          `json:"starred"`
	Hidden    bool          `json:"hidden,omitempty"`
	Links     []domain.Link `json:"links"`
	Scheduled bool          `json:"scheduled"`
	CreatedAt string        `json:"createdAt"`
	UpdatedAt string        `json:"updatedAt"`
}

func toNoteJSON(n store.Note, links []domain.Link) noteJSON {
	if links == nil {
		links = []domain.Link{}
	}
	return noteJSON{
		ID: n.ID, Title: n.Title, BodyMD: n.BodyMD, AuthorID: n.AuthorID,
		X: n.X, Y: n.Y, Color: n.Color, RegionID: optional(n.RegionID),
		Hidden: n.Hidden(), Links: links, Voters: []string{}, CreatedAt: n.CreatedAt, UpdatedAt: n.UpdatedAt,
	}
}

func optional(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

type eventJSON struct {
	ID                string `json:"id"`
	Name              string `json:"name"`
	Lifecycle         string `json:"lifecycle"`
	VotingOpen        bool   `json:"votingOpen"`
	VotesPerUser      int    `json:"votesPerUser"`
	ScheduleThreshold int    `json:"scheduleThreshold"`
}

type regionJSON struct {
	ID    string  `json:"id"`
	Label string  `json:"label"`
	X     float64 `json:"x"`
	Y     float64 `json:"y"`
	W     float64 `json:"w"`
	H     float64 `json:"h"`
	Color string  `json:"color"`
	Z     int     `json:"z"`
}

type slotJSON struct {
	ID      string `json:"id"`
	StartAt string `json:"startAt"`
	EndAt   string `json:"endAt"`
}

type waveJSON struct {
	ID      string     `json:"id"`
	Name    string     `json:"name"`
	Status  string     `json:"status"`
	OpensAt *string    `json:"opensAt"`
	Tracks  int        `json:"tracks"`
	Slots   []slotJSON `json:"slots"`
}

func toWaveJSON(w store.Wave) waveJSON {
	j := waveJSON{ID: w.ID, Name: w.Name, Status: w.Status, OpensAt: optional(w.OpensAt), Tracks: w.Tracks, Slots: []slotJSON{}}
	for _, sl := range w.Slots {
		j.Slots = append(j.Slots, slotJSON{sl.ID, sl.StartAt, sl.EndAt})
	}
	return j
}

type assignmentJSON struct {
	ID      string  `json:"id"`
	NoteID  string  `json:"noteId"`
	SlotID  string  `json:"slotId"`
	Track   int     `json:"track"`
	MeetURL *string `json:"meetUrl"`
}

func toAssignmentJSON(a store.Assignment) assignmentJSON {
	return assignmentJSON{a.ID, a.NoteID, a.SlotID, a.Track, optional(a.MeetURL)}
}

type messageJSON struct {
	ID        string  `json:"id"`
	AuthorID  string  `json:"authorId"`
	ThreadID  *string `json:"threadId"`
	Body      string  `json:"body"`
	Hidden    bool    `json:"hidden,omitempty"`
	CreatedAt string  `json:"createdAt"`
}

type snapshotJSON struct {
	Event       eventJSON                `json:"event"`
	Users       []userJSON               `json:"users"`
	Notes       []noteJSON               `json:"notes"`
	Regions     []regionJSON             `json:"regions"`
	Waves       []waveJSON               `json:"waves"`
	Assignments []assignmentJSON         `json:"assignments"`
	Messages    map[string][]messageJSON `json:"messages"`
	Me          struct {
		VotesRemaining int `json:"votesRemaining"`
		VotesUsed      int `json:"votesUsed"`
	} `json:"me"`
}

// E is an event body: {"kind": ..., fields...}.
type E map[string]any

func mustJSON(v any) []byte {
	b, err := json.Marshal(v)
	if err != nil {
		panic(err) // only our own plain data types are encoded
	}
	return b
}

func helloFrame(u store.User, seq int64) []byte {
	return mustJSON(map[string]any{"type": "hello", "you": toUserJSON(u), "eventSeq": seq})
}

func snapshotFrame(seq int64, s snapshotJSON) []byte {
	return mustJSON(map[string]any{"type": "snapshot", "seq": seq, "state": s})
}

func eventFrame(seq int64, ev E) []byte {
	return mustJSON(map[string]any{"type": "event", "seq": seq, "event": ev})
}

func ackFrame(cmdID string, seq int64) []byte {
	return mustJSON(map[string]any{"type": "ack", "cmdId": cmdID, "seq": seq})
}

func errorFrame(cmdID string, e *domain.CmdError) []byte {
	return mustJSON(map[string]any{"type": "error", "cmdId": cmdID, "code": e.Code, "message": e.Message})
}

var pongFrame = []byte(`{"type":"pong"}`)

File: web/src/api/protocol.ts
// WebSocket protocol types (SPEC §8). Mirrors internal/server/protocol.go.

import type { Role } from './http';

export type Lifecycle = 'setup' | 'active' | 'done';
export type NoteColor = 'yellow' | 'pink' | 'blue' | 'green' | 'orange' | 'purple';
export type LinkKind = 'doc' | 'slides' | 'other';

export const NOTE_COLORS: NoteColor[] = ['yellow', 'pink', 'blue', 'green', 'orange', 'purple'];
export const NOTE_W = 180;
export const NOTE_H = 120;
/** Board grid (and snap) step, in board units. */
export const GRID = 20;
/** Smallest region the server accepts, in board units. */
export const MIN_REGION = 100;

export interface User {
  id: string;
  name: string;
  role: Role;
}

export interface Link {
  id: string;
  title: string;
  url: string;
  kind: LinkKind;
}

export interface Note {
  id: string;
  title: string;
  bodyMd: string;
  authorId: string;
  x: number;
  y: number;
  color: NoteColor;
  regionId: string | null;
  voteTotal: number;
  /** Whether this user voted for it (one vote per person per note). */
  voted: boolean;
  starred: boolean;
  hidden?: boolean;
  links: Link[];
  scheduled: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface EventInfo {
  id: string;
  name: string;
  lifecycle: Lifecycle;
  votingOpen: boolean;
  votesPerUser: number;
}

export interface Region {
  id: string;
  label: string;
  x: number;
  y: number;
  w: number;
  h: number;
  color: string;
  z: number;
}

export interface Slot {
  id: string;
  startAt: string;
  endAt: string;
}

export interface Wave {
  id: string;
  name: string;
  status: 'planned' | 'open' | 'locked' | 'done';
  opensAt: string | null;
  slots: Slot[];
}

export interface Room {
  id: string;
  name: string;
  meetUrl: string | null;
}

export interface Assignment {
  id: string;
  noteId: string;
  slotId: string;
  roomId: string;
}

export interface Message {
  id: string;
  authorId: string;
  threadId: string | null;
  body: string;
  hidden?: boolean;
  createdAt: string;
}

export interface Snapshot {
  event: EventInfo;
  users: User[];
  notes: Note[];
  regions: Region[];
  waves: Wave[];
  rooms: Room[];
  assignments: Assignment[];
  messages: Record<string, Message[]>;
  me: { votesRemaining: number; votesUsed: number };
}

/** Events the client applies. Unknown kinds (from later milestones) are ignored. */
export type BoardEvent =
  | { kind: 'note_created'; note: Note }
  | { kind: 'note_updated'; noteId: string; title: string; bodyMd: string; color: NoteColor; updatedAt: string }
  | { kind: 'note_moved'; noteId: string; x: number; y: number; byUserId: string }
  | { kind: 'note_deleted'; noteId: string }
  | { kind: 'links_set'; noteId: string; links: Link[] }
  | { kind: 'user_joined'; user: User }
  | { kind: 'role_set'; userId: string; role: Role }
  | { kind: 'lifecycle_set'; lifecycle: Lifecycle }
  | { kind: 'region_created'; region: Region }
  | { kind: 'region_updated'; region: Region }
  | { kind: 'region_deleted'; regionId: string }
  | { kind: 'note_retagged'; noteId: string; regionId: string | null }
  | { kind: 'note_starred'; noteId: string }
  | { kind: 'note_unstarred'; noteId: string }
  | { kind: 'vote_cast'; noteId: string; byUserId: string; total: number }
  | { kind: 'vote_retracted'; noteId: string; byUserId: string; total: number }
  | { kind: 'voting_set'; open: boolean }
  | { kind: 'votes_per_user_set'; n: number };

export type ServerFrame =
  | { type: 'hello'; you: User; eventSeq: number }
  | { type: 'snapshot'; seq: number; state: Snapshot }
  | { type: 'event'; seq: number; event: BoardEvent | { kind: string } }
  | { type: 'ack'; cmdId: string; seq: number }
  | { type: 'error'; cmdId: string; code: ErrorCode; message: string }
  | { type: 'pong' };

export type ErrorCode = 'bad_request' | 'forbidden' | 'not_allowed_now' | 'rate_limited' | 'internal';

/** Command payloads by name (SPEC §8.2) — the subset implemented so far. */
export interface Commands {
  create_note: { title: string; bodyMd?: string; x: number; y: number; color?: NoteColor };
  update_note: { noteId: string; title?: string; bodyMd?: string; color?: NoteColor };
  move_note: { noteId: string; x: number; y: number };
  delete_note: { noteId: string };
  set_links: { noteId: string; links: { title: string; url: string; kind: LinkKind }[] };
  set_lifecycle: { lifecycle: Lifecycle };
  create_region: { label: string; x: number; y: number; w: number; h: number; color: string; z?: number };
  update_region: {
    regionId: string;
    label?: string;
    x?: number;
    y?: number;
    w?: number;
    h?: number;
    color?: string;
    z?: number;
  };
  delete_region: { regionId: string };
  star_note: { noteId: string };
  unstar_note: { noteId: string };
  cast_vote: { noteId: string };
  retract_vote: { noteId: string };
  set_voting: { open: boolean };
  set_votes_per_user: { n: number };
}
export type CommandName = keyof Commands;
